libs/s25main/buildings/nobBaseWarehouse.cpp

```diff
@@ -1026,15 +1026,14 @@ std::unique_ptr<nofDefender> nobBaseWarehouse::ProvideDefender(nofAttacker& atta
     // Kein Soldat gefunden, als letzten Hoffnung die Soldaten nehmen, die ggf in der Warteschlange noch hängen
     for(auto it = leave_house.begin(); it != leave_house.end(); ++it)
     {
-        nofSoldier* soldier;
+        std::unique_ptr<nofSoldier> soldier;
         // Soldat?
         if((*it)->GetGOT() == GO_Type::NofAggressivedefender)
         {
-            auto* aggDefender = static_cast<nofAggressiveDefender*>(it->get());
-            aggDefender->NeedForHomeDefence();
-            soldier = aggDefender;
+            soldier.reset(static_cast<nofSoldier*>(it->release()));
+            static_cast<nofAggressiveDefender&>(*soldier).NeedForHomeDefence();
         } else if((*it)->GetGOT() == GO_Type::NofPassivesoldier)
-            soldier = static_cast<nofPassiveSoldier*>(it->get());
+            soldier.reset(static_cast<nofSoldier*>(it->release()));
```

Comment:
and here couldn't we simply write something like
```suggestion
soldier = std::dynamic_pointer_cast<nofSoldier>(*it));
```

Reply:
Because a dynamic cast is to much here. We already pay the virtual cost by the GetGOT call above, so no need to pay it again.
However I'm thinking about whether we should add our own methods for dynamic_pointer_cast of unique ptrs (yes we have one already, but that is destructive in case of a type mismatch) and static_pointer_cast (that can be destructive in all cases)

Reply:
Done